src/core/lib/transport/promise_endpoint.h

```diff
@@ -0,0 +1,282 @@
+                          &pending_read_buffer_,
+                          nullptr /* uses default arguments */)) {
+        ReadCallback(absl::OkStatus(), num_bytes, absl::nullopt);
+      }
+    } else {
+      read_result_ = absl::OkStatus();
+    }
+    return [this, num_bytes]() -> Poll<absl::StatusOr<SliceBuffer>> {
+      MutexLock lock(&read_mutex_);
+      if (!read_result_.has_value()) {
+        read_waker_ = Activity::current()->MakeNonOwningWaker();
+        return Pending();
+      } else if (!read_result_->ok()) {
+        const absl::Status ret = *read_result_;
+        read_result_.reset();
+        return ret;
+      } else {
+        SliceBuffer ret;
+        grpc_slice_buffer_move_first(read_buffer_.c_slice_buffer(), num_bytes,
+                                     ret.c_slice_buffer());
+        read_result_.reset();
+        return std::move(ret);
+      }
+    };
+  }
+
+  // Returns a promise that resolves to `Slice` with at least
+  // `num_bytes` bytes which should be less than INT64_MAX bytes.
+  //
+  // Concurrent reads are not supported, which means callers should not call
+  // `ReadSlice()` before the previous read finishes. Doing that results in
+  // undefined behavior.
+  auto ReadSlice(size_t num_bytes) {
+    ReleasableMutexLock lock(&read_mutex_);
+    // Previous read result has not been polled.
+    GPR_ASSERT(!read_result_.has_value());
+    // Should not have pending reads.
+    GPR_ASSERT(pending_read_buffer_.Count() == 0u);
+    if (read_buffer_.Length() < num_bytes) {
+      lock.Release();
+      const struct grpc_event_engine::experimental::EventEngine::Endpoint::
+          ReadArgs read_args = {static_cast<int64_t>(num_bytes)};
+
+      // If `Read()` returns true immediately, the callback will not be
+      // called. We still need to call our callback to pick up the result
+      // and maybe do further reads.
+      if (endpoint_->Read(
+              std::bind(&PromiseEndpoint::ReadCallback, this,
+                        std::placeholders::_1, num_bytes, read_args),
+              &pending_read_buffer_, &read_args)) {
+        ReadCallback(absl::OkStatus(), num_bytes, read_args);
+      }
+    } else {
+      read_result_ = absl::OkStatus();
+    }
+    return [this, num_bytes]() -> Poll<absl::StatusOr<Slice>> {
+      MutexLock lock(&read_mutex_);
+      if (!read_result_.has_value()) {
+        read_waker_ = Activity::current()->MakeNonOwningWaker();
+        return Pending();
+      } else if (!read_result_->ok()) {
+        const auto ret = *read_result_;
+        read_result_.reset();
+        return ret;
+      } else if (read_buffer_.RefSlice(0).size() == num_bytes) {
+        read_result_.reset();
+        return Slice(read_buffer_.TakeFirst().TakeCSlice());
+      } else {
+        MutableSlice ret = MutableSlice::CreateUninitialized(num_bytes);
+        read_buffer_.MoveFirstNBytesIntoBuffer(num_bytes, ret.data());
```

Comment:
Probably for the future: This will do a malloc and memcpy to copy the `num_bytes` into a slice. Depending on where this ReadSlice method is going to be invoked (for e.g in the critical path), we should think about some optimization to avoid the memcpy

Reply:
Thanks for catching this, I added a TODO comment.
So If the `read_buffer_ `slice size is equal to the requested `num_bytes`, then there won't be a memcpy. But if the `read_buffer_` slice size is not equal to the requested `num_bytes`, it needs to copy the buffer to a slice that has size equal to `num_bytes`. 

Is there an existing function that could return any size Slice from the SliceBuffer based on the start location? If not, I think we will need to add one in SliceBuffer to avoid this memcpy.

Reply:
I dont think there is an easy way to do this based on the current grpc_slice_buffer and grpc_slice implementations. Lets try to avoid using this method in the critical sections and if that is not possible, we could discuss some alternatives.

Also nit: I think you also need to handle the case where read_buffer_.Size() < num_bytes. Its not always the case that EventEngine endpoints will return at least num_bytes.

Reply:
I see, we could use `Read(num_bytes)` in the transport implement to avoid memcpy. 

>I think you also need to handle the case where read_buffer_.Size() < num_bytes. Its not always the case that EventEngine endpoints will return at least num_bytes.

Yes, there's a `GPR_ASSERT(src->length >= n);` check in `MoveFirstNBytesIntoBuffer` method. Or should I handle `read_buffer_.Size() < num_bytes` case in Promise Endpoint separately? I was expecting it to throw an assert failed error with `read_buffer_.Size() < num_bytes`, but if it isn't the best practice, I could try to return a `absl::Status(absl::StatusCode::kInternal, absl::StrFormat( "EventEngine read buffer size %d is less than the request bytes %d.", read_buffer_.Size(), num_bytes))`

Reply:
I think you'd need to handle the read_buffer_.Size() < num_bytes case similar to how it is handled in the Read(num_bytes) method. It is not guaranteed that the event engine endpoint will always return at least num_bytes amount of data. Some event engines may ignore the ReadArgs passed to the endpoint (e.g, windows, ios).

Reply:
I see, actually `ReadSlice(num_bytes)` is handling `read_buffer_.Size() < num_bytes` same as the `Read(num_bytes)`. It first checks if `read_buffer_.Size() < num_bytes`, if true then do event engine endpoint::Read, and then schedule the `ReadCallback`, and continuously check if `read_buffer_.Size() < num_bytes` in `ReadCallback`.  So at the end `read_buffer_.Size()` is expected to be greater or equal to num_bytes.

Reply:
Oh I missed it. thanks for the clarification.